crates/sargon/src/factor_instances_provider/provider/outcome/internal_factor_instances_provider_outcome.rs

```diff
@@ -2,13 +2,21 @@ use crate::prelude::*;
 
 #[derive(Clone, Debug)]
 pub struct InternalFactorInstancesProviderOutcome {
+    pub per_derivation_preset: IndexMap<
```

Comment:
Hmm... should it be:
* PER FactorSource PER DerivationPreset or
* PER DerivationPreset PER FactorSource
?
I think the latter is better, because it would allow to query the FIP with different FactorSources per DerivationPreset, which would allow e.g. `ShieldA` to be used for personas `X` & `Y` but `ShieldB` (where `ShieldA != ShieldB`) for account `A` & `B` **to be queried for in the same single pass to FIP**. So that is one argument in favour of `PER DerivationPreset PER FactorSource`. Are there any arguments in favour of `PER FactorSource PER DerivationPreset` or?

@matiasbzurovski @GhenadieVP I would appriciated any early input here!

Reply:
(I've begun upgrading the `FactorInstancesCache` first, which I think is near complete)

The `InternalFactorInstancesProviderOutcome` is ofc used by `FactorInstancesProviderOutcome` (which I have not updated yet).

So how are there outcomes used? Well they are used by `VirtualEntityCreatingInstanceProvider` and then in turn by `Profile` method [`create_unsaved_entities_with_factor_source_with_derivation_outcome `](https://github.com/radixdlt/sargon/blob/7f1368319f29e93024635d7572f9fd6133e2cad2/crates/sargon/src/profile/logic/create_entity.rs#L35-L67) which will essentially remain the same, since it will NOT query the FIP (via the`VirtualEntityCreatingInstanceProvider` ) with multiple DerivationPresets, but rather just the one (either `DerivationPreset::AccountVeci` or `DerivationPresets::IdentityVeci`).

So lets look at the use case where we will query the FIP with multiple derivation presets:

Reply:
**Gated behind `cfg(test)`: `make_security_structure_of_factor_instances_for_entities_without_consuming_cache_with_derivation_outcome`**
which you find here:
https://github.com/radixdlt/sargon/blob/7f1368319f29e93024635d7572f9fd6133e2cad2/crates/sargon/src/system/sargon_os/sargon_os_accounts.rs#L800-L837

It is, as mentioned above gated behind `cfg(test)` so not at all final... and we note that the signature is now:

```rust
    pub(crate) async fn make_security_structure_of_factor_instances_for_entities_without_consuming_cache_with_derivation_outcome<
        A: IsEntityAddress,
    >(
        &self,
        addresses_of_entities: IndexSet<A>,
        security_structure_of_factor_sources: SecurityStructureOfFactorSources, // Aka "shield"
    ) -> Result<(
```

**But this PR WILL change the signature to NOT be generic over `E`**, so `addresses_of_entities: IndexSet<A>` is going to be changed to `IndexSet<AddressOfAccountOrPersona>`.

[So instead of](https://github.com/radixdlt/sargon/blob/7f1368319f29e93024635d7572f9fd6133e2cad2/crates/sargon/src/system/sargon_os/sargon_os_accounts.rs#L852-L854):
```rust
addresses_of_entities.clone().into_iter().map(|entity_address|
```

we probably wanna split the `addresses_of_entities` into Persona addresses and query `outcome.get(DerivationPreset::IdentityMfa)` and then iterate... and then the same for accounts?

Reply:
So the more I write about this the more convinced I become that we want `PER DerivationPreset PER FactorSource`. or rather NEED.

Reply:
Yes `PER DerivationPreset PER FactorSource` makes more sense, as in different contexts we need to use the different preset, and it is nice to be able to slice all of the factor instances per preset.

Reply:
This is new - introduce by this PR:
```rust
#[derive(Debug, PartialEq, Eq)]
pub struct CacheInstancesAndRemainingQuantityToDerive {
    pub instances_to_use_from_cache: FactorInstances, // if empty then this was not a requested derivation preset, but we are cache filling and found `quantity_to_derive` needed to fill cache.
    pub quantity_to_derive: usize,
}
```

So I should probably change this - which is also new in this PR:
```diff
#[derive(Debug, PartialEq, Eq)]
pub struct CacheNotSatisfied { // NEW introduced in this PR
-   /// PER FactorSourceID => PER DerivationPreset => CacheInstancesAndRemainingQuantityToDerive
+   /// PER DerivationPreset => PER FactorSourceID => CacheInstancesAndRemainingQuantityToDerive
    pub cached_and_quantities_to_derive: IndexMap<
-        FactorSourceIDFromHash,
+       DerivationPreset
        IndexMap<
-             DerivationPreset, 
+             FactorSourceIDFromHash, 
             CacheInstancesAndRemainingQuantityToDerive
        >,
    >,
}
```

Reply:
At the cache level we do store per factor source id, right?

Reply:
Yes the cache is:

```rust
pub(super) type FICStorage = IndexMap<
    FactorSourceIDFromHash,
    IndexMap<IndexAgnosticPath, FactorInstances>,
>;
```

But I think that makes sense to keep as is, since, it allows for easy add/delete instances based on new FactorSource.

Reply:
So I think it makes sense to keep that, but ""transposing"" it to:

```rust
#[derive(Debug, PartialEq, Eq)]
pub struct CacheNotSatisfied {
    /// PER DerivationPreset => PER FactorSourceID => CacheInstancesAndRemainingQuantityToDerive
    pub cached_and_quantities_to_derive: IndexMap<
        DerivationPreset,
        IndexMap<
            FactorSourceIDFromHash,
            CacheInstancesAndRemainingQuantityToDerive,
        >,
    >,
}
```

which then quite easily and naturally maps to;

```rust
#[derive(Clone, Debug)]
pub struct InternalFactorInstancesProviderOutcome {
    pub per_derivation_preset: IndexMap<
        DerivationPreset,
        InternalFactorInstancesProviderOutcomePerFactor,
    >,
}
```

which will then be quite easy to use in `make_security_structure_of_factor_instances_for_entities_without_consuming_cache_with_derivation_outcome` which will "group" entities of kind Persona and entities of kind Account together and use the `DerivationPreset::AccountMfa` and `DerivationPreset::IdentityMfa` respectively as "key" to get their `InternalFactorInstancesProviderOutcomePerFactor`, from the outcome of the FIP.

I think that will turn out quite nicely